Dogbreed get

TroyMcl opened the pull request:
To set up a file structure for future project growth I added a services directory to src where various abstracted business functionality can be written. For this challenge, I created 2 subdirectories, one for HTTP requests and one for utility functions.

I decided to abstract out the HTTP request and the code to transform the data into an array of strings from the lambda. I did this in an effort to make the code more reusable and for testing reasons. I find it easier to write unit tests when the code is broken down into units of functionality like this.

The utility function flattenObjectValueKeyOrder is set up to work with generics. The idea was to make it more reusable, so it is not locked into only working with a specific type like the data returned by GET request to the dog breed API. This function is a shallow flattening but it will flatten an object with arrays, strings, numbers, and booleans as values and return an array of strings. The trade-off I made for this flexibility is a higher cyclomatic complexity as it currently has some nested conditional logic.

The API function getDogBreeds is where I decided to handle an external call timing out. I also decided to handle errors locally inside the API function rather than automatically bubble every error back to the client. My reasoning for this was the client doesn't need to see every error, a simple message that something broke is ok in most cases. Also if remote logging is being used that code could be incorporated into the API function preserving the error and letting the lambda handle how to build its response.

TroyMcl committed:
Initial file structure set up
Added some types and an enum for the services functions and the Lambda
Build utility function for converting API response to string array
Created first test for utility function

TroyMcl committed:
Added GET API request for dog breeds
Added test for each possibility of the api function (success, an error from the server, and a timed out request)
Added one more test for the utility flatten function

TroyMcl committed:
Added GET dog breeds lambda
Added test for success and errors in lambda

seanmooney-extend commented on src/lambdas/breeds-get.ts:
```diff
@@ -0,0 +1,22 @@
+      statusCode: 200,
+      body: payload,
+    }
+  } catch (err) {
+    return {
+      statusCode: 500,
+      message: err.message,
+    }
+  }
```
If a timeout were to occur, could we return a more specific response and still handle the generic 500 errors as well?

seanmooney-extend replied:
We also want to be able to keep `err` typed as unknown since we do not know the type of error that was thrown. Is there a way we can ensure the error thrown was related to a timeout?

TroyMcl replied:
Thanks for pointing this out. This implementation is definitely too brittle. I decided to move the logic for generating ErrorResponse objects to a helper function where different error messages and response codes can be set depending on the error name.

seanmooney-extend commented on src/lambdas/breeds-get.ts:
```diff
@@ -0,0 +1,22 @@
+    const apiData: RequestedDogBreeds = await getDogBreeds()
+    if (apiData.status === 'error') {
+      throw new Error('Something went wrong')
+    }
```
Couldn't we throw an error within `getDogBreeds` and bubble up the error to be handled within this try/catch statement?

TroyMcl replied:
Yes you're right, I totally tried to over engineer getDogBreeds and it made the lambda more complex than it should have been. I removed the try catch from getDogBreeds so it will either return data, or result in an error. This code block is gone now

seanmooney-extend commented on src/services/utilities/flattenObjectValueKeyOrder.ts:
```diff
@@ -0,0 +1,17 @@
+export const flattenObjectValueKeyOrder = <T>(obj: T): string[] => {
+  return Object.keys(obj).reduce((acc, key) => {
+    const value = obj[key as keyof T]
+
+    if (Array.isArray(value)) {
+      if (value.length < 1) {
+        acc.push(key)
+      } else {
+        const subValues = value.map((item) => `${item} ${key}`)
+        acc.push(...subValues)
+      }
+    } else {
+      acc.push(`${value} ${key}`)
+    }
+    return acc
+  }, [] as string[])
+}
```
This is good 👍 I like that you broke this out into it's own function

TroyMcl replied:
thank you

seanmooney-extend commented on src/services/http-req/dogBreedRequest.ts:
```diff
@@ -0,0 +1,32 @@
+      status: Status.success,
+      data: payload,
+    }
+  } catch (err: any) {
```
Does this catch statement seem necessary since it's a layer below the handler? I'm curious if it provides any value. If it does we should keep `err` typed as unknown here as well too 🙂

TroyMcl replied:
I don't think the try catch is necessary in this function and its actually making the code harder to understand and work with. I removed those blocks and the function now returns either the api data or throws an error for the function invoking this one to handle. I did change the type of err in breed-get.ts to unknown though 😅

seanmooney-extend replied:
Awesome, thank you for doing that 👍

seanmooney-extend commented on src/services/http-req/dogBreedRequest.ts:
```diff
@@ -0,0 +1,32 @@
+
+    if (reqTimedOut) {
+      throw new Error('Request to externial API timed out')
+    }
```
I believe when `abort()` is called it throws an error, I'm not sure the `reqTimedOut` is necessary 🙂

TroyMcl replied:
I saw that too but I could not get it to work in the code for some reason. I also tried the node-abort-controller package which is supposed to be a polyfill for node. I am not sure what I am missing, because I tried adding the example code from node-fetch and node-abort-controller and running it but I still couldn't get an error either in the function or in the test suite running it. I guess this is my work around right now, but if you see something I missed I would really like to be able to get this to work as expected.

seanmooney-extend replied:
I left some comments around this, hopefully that points you in the right direction. If you're still not getting it to work you can reach out here and I can answer any questions you might have

TroyMcl replied:
thanks, yes those comments were the aha moment of what I was missing

seanmooney-extend commented on src/lambdas/mock-events/default.ts:
```diff
@@ -1 +1,23 @@
+  'japanese spaniel',
+  'sussex spaniel',
+  'welsh spaniel',
+]
```
This data is repeated in the `mock-values` file, just checking if this was intended?

TroyMcl replied:
🤦‍♂️  thanks for pointing that out. I forgot to clear out this file after trying something with it. Total oversite on my part, updated with the latest push

seanmooney-extend replied:
No worries!

seanmooney-extend reviewed (commented):
Hi Troy 👋 thank you for taking the time to work on this. 

I like that you're using the `AbortController` in your implementation to handle the timeout, that's definitely a good approach. I left some comments below around the general implementation, feel free to reach out with any questions and I'll be happy to answer them.

Best,

Sean

TroyMcl committed:
Removed mock values from default.ts that are located in mock-values.ts

TroyMcl committed:
Refactored API function to simplify logic:
Removed internal Async Await inside the code. The function will always be called in an async context and API Errors should be handled in that context

TroyMcl committed:
Updated RequestAPIData type to match refactor of API function

TroyMcl committed:
Refactored and simplified Lambda:
API function is not returing a custom object, so the check for an error in the custome object was removed
Abstracted out generating the error response object to a new function where different error messages can be returned based on the error

seanmooney-extend commented on src/services/http-req/dogBreedRequest.test.ts:
```diff
@@ -0,0 +1,71 @@
+    expect(response instanceof Error).toBe(true)
+    expect(response.message).toEqual('I broke somewhere')
+  })
+})
```
Typically we don't want to use a try catch inside of our test suite. Instead is there a way we can expect the promise to reject?

TroyMcl replied:
OK, that makes sense. I adjusted these tests to expect a rejection that equals an error. I do have a question about asserting errors. Is it better to run assertions on the error's name and message properties, or to run the assertion on the entire error?

seanmooney-extend replied:
Typically we like to test the whole error, we try to keep the assertions to a minimum so something like this would be good:

`expect(promise).rejects.toThrowError(new Error('whoopsie'))`

It tests the message, the type, and everything in between, all in one assertion 👍 but no worries, what you have is good!

TroyMcl replied:
Thank you for clarifying this for me it makes a lot of sense now

seanmooney-extend commented on src/services/http-req/dogBreedRequest.test.ts:
```diff
@@ -0,0 +1,71 @@
+    mockedFetch.mockReturnValueOnce({
+      json: () => {
+        return new Promise((resolve) => {
+          setTimeout(() => {
+            resolve({ this: 'should not matter' })
+          }, 3000)
```
I'm not sure this `setTimeout` has any value here, we are trying to mock a return value and not an implementation. Instead we might get more value out of returning the error we're expecting from fetch and checking that our code works off that. This way we can also manipulate the error to match something that an abort controller would throw 🙂

seanmooney-extend replied:
I see what you're doing here though, the timeout in the code is 1500 and you're saying this should wait 3000, that's pretty clever but there are cleaner ways to accomplish this 👍

TroyMcl replied:
I think this comment was my aha moment about what I was doing wrong, thank you. Since the API call is being mocked we won't see an actual AbortError from the abort controller since the listener attached to the fetch isn't there. I believe this means the test doesn't need to worry about a delay, it just needs to make sure any errors that occur in this function are passed on correctly.

seanmooney-extend commented on src/services/http-req/dogBreedRequest.ts:
```diff
@@ -0,0 +1,24 @@
+
+export async function getDogBreeds(): Promise<RequestedDogBreeds> {
+  const controller = new AbortController()
+  let reqTimedOut = false
+  const timeout = setTimeout(() => {
+    controller.abort()
+    reqTimedOut = true
+  }, 1500)
+
+  const res = await fetch('https://dog.ceo/api/breeds/list/all', { signal: controller.signal })
+  const payload = await res.json()
+  clearTimeout(timeout)
+
+  if (reqTimedOut) {
+    const error = new Error('Request to externial API timed out')
+    error.name = 'RequestTimeOut'
+    throw error
+  }
```
Your abort controller implementation looks right to me, I would check my comments on your test suite for more info. You're definitely on the right track! If we can remove the need for `reqTimedOut` I think this would be solid.

TroyMcl replied:
haha yeah the comments in the test suite cleared this up for me, thanks

seanmooney-extend commented on src/services/gererateErrorResponse.ts:
```diff
@@ -0,0 +1,15 @@
+import { ErrorResponse } from '../lambdas/types'
+
+export const generateErrorResponse = (error: unknown): ErrorResponse => {
+  const response = {
+    statusCode: 500,
+    message: 'Something went wrong',
+  }
+  if (error instanceof Error) {
+    if (error.name === 'RequestTimeOut') {
+      response.statusCode = 408
+      response.message = 'Request to externial API timed out'
+    }
+  }
+  return response
+}
```
This is excellent error handling 👍 

Two things worth mentioning though:
- We can just return from within the if statement so we don't need to reassign values on the constant
- The error name would change if this were an actual abort error, I think once the code from my other feedback is in place this would need to change 🙂

TroyMcl replied:
I updated the condition looking at error.name to look for AbortError instead of my made-up magic name 😬 . I removed the response const and just have objects being directly returned in the if statement and as a default.

seanmooney-extend commented on src/lambdas/breeds-get.test.ts:
```diff
@@ -0,0 +1,44 @@
+  it('should have a correctly formated body array', async () => {
+    const response: any = await getBreeds()
+    expect(response.body.length).toEqual(assert1.length)
+    expect(response.body.join(' ')).toEqual(assert1.join(' '))
```
I'm a little curious about the `join(' ')` on both the body and the mock value. Without the join wouldn't these be equal as well?

TroyMcl replied:
LOL, I have the habit of doing this when comparing arrays (turning them into strings). Completely not needed here so it has been removed.

seanmooney-extend commented:
Apparently GitHub publishes a review when I reply to comments, sorry about that!

I left a few more comments but the PR is definitely looking good. As always feel free to let me know if you have any questions.

Best,

Sean

TroyMcl committed:
Updated RequestedAPIData type as the option of string was not necessary

TroyMcl committed:
Removed the reqTimedOut variable and logic from the API call. Removed async/await from the test suite.

TroyMcl committed:
generateErrorResponse returns directly from the if block and has a default. Updated if condition and test suite to check for AbortError that I am expecting abort-controller to throw
